AdblockPlusSafari/AdblockPlusExtras.m

```diff
@@ -0,0 +1,326 @@
+            filterlist[@"updating"] = @NO;
+            filterlists[key] = filterlist;
+          }
+          sSelf.filterLists = filterlists;
+        }
+      }
+    }];
+  }
+  return self;
+}
+
+#pragma mark - property
+
+@dynamic lastUpdate;
+@dynamic updating;
+
+- (NSDate *)lastUpdate
+{
+  return [[self.filterLists allValues] valueForKeyPath:@"@min.lastUpdate"];
+}
+
+- (BOOL)updating
+{
+  return [[[self.filterLists allValues] valueForKeyPath:@"@sum.updating"] integerValue] > 0;
+}
+
+- (void)setFilterLists:(NSDictionary<NSString *,NSDictionary<NSString *,NSObject *> *> *)filterLists
+{
+  BOOL wasUpdating = self.updating;
+  BOOL hasAnyLastUpdateFailed = self.anyLastUpdateFailed;
+
+  [self willChangeValueForKey:@"lastUpdate"];
+  [self willChangeValueForKey:@"updating"];
+  super.filterLists = filterLists;
+  [self didChangeValueForKey:@"updating"];
+  [self didChangeValueForKey:@"lastUpdate"];
+
+  BOOL updating = self.updating;
+  BOOL anyLastUpdateFailed = self.anyLastUpdateFailed;
+
+  if (self.installedVersion < self.downloadedVersion && wasUpdating && !updating) {
+    // Force content blocker to load newer version of filterlist
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+
+  if (hasAnyLastUpdateFailed != anyLastUpdateFailed) {
+    self.needsDisplayErrorDialog = anyLastUpdateFailed;
+    [self displayErrorDialogIfNeeded];
+  }
+}
+
+- (void)setNeedsDisplayErrorDialog:(BOOL)needsDisplayErrorDialog
+{
+  _needsDisplayErrorDialog = needsDisplayErrorDialog;
+  [self.adblockPlusDetails setBool:needsDisplayErrorDialog forKey:AdblockPlusNeedsDisplayErrorDialog];
+  [self.adblockPlusDetails synchronize];
+}
+
+- (BOOL)anyLastUpdateFailed
+{
+  return [[[self.filterLists allValues] valueForKeyPath:@"@sum.lastUpdateFailed"] integerValue] > 0;
+}
+
+#pragma mark -
+
+- (void)setEnabled:(BOOL)enabled reload:(BOOL)reload
+{
+  self.enabled = enabled;
+
+  if (reload) {
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+- (void)setAcceptableAdsEnabled:(BOOL)enabled reload:(BOOL)reload
+{
+  self.acceptableAdsEnabled = enabled;
+
+  if (reload) {
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+- (void)reloadContentBlockerWithCompletion:(void(^__nullable)(NSError * __nullable error))completion;
+{
+  __weak __typeof(self) wSelf = self;
+  wSelf.reloading = YES;
+  [SFContentBlockerManager reloadContentBlockerWithIdentifier:self.contentBlockerIdentifier completionHandler:^(NSError *error) {
+    NSLog(@"%@", error);
+    dispatch_async(dispatch_get_main_queue(), ^{
+      wSelf.reloading = NO;
+      [wSelf checkActivatedFlag];
+      if (completion) {
+        completion(error);
+      }
+    });
+  }];
+}
+
+- (void)checkActivatedFlag
+{
+  BOOL activated = [self.adblockPlusDetails boolForKey:AdblockPlusActivated];
+  if (self.activated != activated) {
+    self.activated = activated;
+  }
+}
+
+- (void)updateFilterlists:(BOOL)userTriggered
```

Comment:
userTriggered flag is used just for different error messages. It does not have more use.